components/core/src/package/install.rs

```diff
@@ -318,35 +375,124 @@ impl PackageInstall {
+        let mut legacy_run_paths: Vec<PathBuf> = Vec::new();
+        let mut has_legacy_run_paths = false;
+
+        let env = self.environment()?;
+        if !env.is_empty() {
+            if let Some(path) = env.get("PATH") {
+                let mut v: Vec<PathBuf> =
+                    env::split_paths(&path).map(|p| PathBuf::from(&p)).collect();
+                legacy_run_paths.append(&mut v);
+            }
+
+            for (key, value) in env.into_iter() {
+                run_envs.insert(key, value);
+            }
+        } else {
+            let mut p = self.paths()?;
+            legacy_run_paths.append(&mut p);
+            has_legacy_run_paths = true;
+        }
 
-        let mut p = try!(self.paths());
-        run_paths.append(&mut p);
-        idents.insert(self.ident().clone());
-        let deps: Vec<PackageInstall> = try!(self.load_deps());
+        let deps = self.load_deps()?;
         for dep in deps.iter() {
-            let mut p = try!(dep.paths());
-            run_paths.append(&mut p);
+            let env = dep.environment()?;
+            let env_sep = dep.environment_sep()?;
+            if !env.is_empty() {
+                if let Some(path) = env.get("PATH") {
+                    let mut v: Vec<PathBuf> =
+                        env::split_paths(&path).map(|p| PathBuf::from(&p)).collect();
+                    legacy_run_paths.append(&mut v);
+                }
+
+                for (key, value) in env.into_iter() {
+                    match run_envs.entry(key) {
+                        Occupied(entry) => {
+                            match env_sep.get(entry.key()) {
+                                Some(sep) => {
+                                    let v = entry.into_mut();
+                                    v.push_str(sep);
+                                    v.push_str(&value);
+                                }
+                                None => warn!("Cannot join {}, no separator defined", entry.key()),
+                            }
+                        }
+                        Vacant(entry) => {
+                            entry.insert(value);
+                            ()
+                        }
+                    }
+                }
+            } else {
+                let mut p = dep.paths()?;
+                legacy_run_paths.append(&mut p);
+                has_legacy_run_paths = true;
+            }
             idents.insert(dep.ident().clone());
         }
-        let tdeps: Vec<PackageInstall> = try!(self.load_tdeps());
+
+        let tdeps = self.load_tdeps()?;
         for dep in tdeps.iter() {
             if idents.contains(dep.ident()) {
                 continue;
             }
-            let mut p = try!(dep.paths());
-            run_paths.append(&mut p);
+            let env = dep.environment()?;
+            let env_sep = dep.environment_sep()?;
+            if !env.is_empty() {
```

Comment:
this code block looks similar to https://github.com/habitat-sh/habitat/pull/1830/files#diff-b9b27ac18067cb62ebad852e9abc0b10R413

Perhaps make this DRY?